system/jlib/jmutex.hpp

```diff
@@ -1022,24 +1022,46 @@ class Singleton
         return value;
     }
 
+protected:
+    std::atomic<X *> singleton = {nullptr};
+    std::atomic<bool> initialized = {false};
+    CriticalSection cs;
+};
+
+template <typename X>
+class Singleton : public SingletonBase<X>
+{
+    typedef SingletonBase<X> PARENT;
+public:
+    using PARENT::query;
     //destroy() is designed to be called from a static destructor, not thread safe with calls to query() ...
     void destroy(std::function<void (X*)> destructor)
     {
-        X * value = singleton.exchange(nullptr, std::memory_order_acq_rel);
+        X * value = PARENT::singleton.exchange(nullptr, std::memory_order_acq_rel);
         if (value)
             destructor(value);
-        initialized = false;
+        PARENT::initialized = false;
     }
     void destroy()
     {
-        X * value = singleton.exchange(nullptr, std::memory_order_acq_rel);
+        X * value = PARENT::singleton.exchange(nullptr, std::memory_order_acq_rel);
         delete value;
-        initialized = false;
+        PARENT::initialized = false;
+    }
+};
+
+template <typename X>
+class OwnedSingleton : public SingletonBase<X>
+{
+    typedef SingletonBase<X> PARENT;
+public:
+    using PARENT::query;
+    ~OwnedSingleton()
+    {
+        X * value = PARENT::singleton.exchange(nullptr, std::memory_order_acq_rel);
```

Comment:
This does not need to use exchange (which is relatively expensive), because by definition only a single thread can be accessing it at this point.